src/lib.rs

```diff
@@ -7,292 +7,314 @@
 //! NStack
 //!
 //! A stack datastructure with indexed lookup.
-#![warn(missing_docs)]
-use std::borrow::Borrow;
-use std::marker::PhantomData;
-use std::{io, mem};
+use core::mem;
 
-use kelvin::annotations::{Cardinality, Counter, Nth};
-use kelvin::{
-    Annotation, Branch, BranchMut, ByteHash, Compound, Content, Handle,
-    HandleMut, HandleType, Sink, Source,
-};
+use canonical::{Canon, Store};
+use canonical_derive::Canon;
 
-const N: usize = 4;
+use microkelvin::{Annotated, Annotation, Child, ChildMut, Compound};
 
-/// A stack datastructure with indexed lookup.
-#[derive(Clone)]
-pub struct NStack<T, A, H>([Handle<Self, H>; N], PhantomData<(T, A)>)
-where
-    T: Content<H>,
-    Self: Compound<H>,
-    H: ByteHash;
+const N: usize = 4;
 
-impl<T, A, H> Default for NStack<T, A, H>
+#[derive(Clone, Canon, Debug)]
+pub enum NStack<T, A, S>
 where
-    T: Content<H>,
-    A: Content<H> + Annotation<T, H>,
-    H: ByteHash,
+    A: Canon<S> + Annotation<T>,
+    T: Canon<S>,
+    S: Store,
 {
-    fn default() -> Self {
-        let handles: [Handle<Self, H>; N] = Default::default();
-        NStack(handles, PhantomData)
-    }
+    Leaf([Option<T>; N]),
+    Node([Option<Annotated<NStack<T, A, S>, S>>; N]),
 }
 
-impl<T, A, H> Content<H> for NStack<T, A, H>
+impl<T, A, S> Compound<S> for NStack<T, A, S>
 where
-    T: Content<H>,
-    A: Content<H> + Annotation<T, H>,
-    H: ByteHash,
+    T: Canon<S>,
+    A: Canon<S> + Annotation<T>,
+    S: Store,
 {
-    fn persist(&mut self, sink: &mut Sink<H>) -> io::Result<()> {
-        for handle in self.0.iter_mut() {
-            handle.persist(sink)?;
+    type Leaf = T;
+    type Annotation = A;
+
+    fn annotation(&self) -> Self::Annotation {
+        match self {
+            NStack::Leaf([None, None, None, None]) => A::identity(),
```

Comment:
We are assuming a constant arity, and it should be defined by `hades252::WIDTH`

Reply:
I disagree, this PR is about porting nstack to microkelvin, it's not about changing/discussing the arity or adding dependencies.

Reply:
Why do we have 4 `None`?